File: js/src/outbound_policy.ts
/**
 * Immutable configuration for replacing headers in outbound HTTPS traffic.
 */

import {
  OutboundPolicy as OutboundPolicyProto,
  OutboundPolicy_HeaderReplacement,
} from "../proto/modal_proto/api";
import { InvalidError } from "./errors";
import { Secret, secretEnvDictHydrator } from "./secret";

const DOMAIN_RE =
  /^(\*\.)?([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$/;
const HEADER_NAME_RE = /^[A-Za-z0-9!#$%&'*+.^_`|~-]+$/;
const MAX_HEADER_REPLACEMENTS = 25;

function hasInvalidHeaderValueChar(value: string): boolean {
  for (const c of value) {
    const code = c.codePointAt(0)!;
    if ((code < 0x20 && c !== "\t") || code === 0x7f) {
      return true;
    }
  }
  return false;
}

/**
 * Whether a header value template references a `$KEY`.
 * @internal
 */
export function templateReferencesKey(template: string): boolean {
  const isKeyStart = (c: string) => /[a-zA-Z_]/.test(c);
  let i = template.indexOf("$");
  while (i !== -1 && i + 1 < template.length) {
    const c = template[i + 1];
    if (c === "$") {
      i = template.indexOf("$", i + 2);
      continue;
    }
    if (isKeyStart(c)) {
      return true;
    }
    i = template.indexOf("$", i + 1);
  }
  return false;
}

/** @internal */
export type HeaderReplacement = {
  domain: string;
  headers: Record<string, string>;
  secret?: Secret;
};

/**
 * Immutable configuration for replacing headers in outbound HTTPS requests
 * from a Sandbox.
 *
 * EXPERIMENTAL: the API is subject to change.
 *
 * Header values support templating with keys in a replacement's secret: a
 * `$`-prefixed key name in the secret is replaced with the secret value.
 * Literal `$` characters are written `$$`.
 *
 * Secret values never enter the Sandbox: they are resolved and injected into
 * matching requests outside the container.
 *
 * ```ts
 * const secret = await modal.secrets.fromName("api-token");
 *
 * const outboundPolicy = new modal.ExperimentalOutboundPolicy()
 *   // Inject a secret-backed Authorization header.
 *   .withHeaderReplacement({
 *     domain: "example.com",
 *     secret,
 *     headers: { Authorization: "Bearer $API_TOKEN" },
 *   })
 *   // Inject a static header into requests to another domain.
 *   .withHeaderReplacement({
 *     domain: "modal.com",
 *     headers: { "X-Trace-Token": "trace_abcd" },
 *   });
 *
 * const sb = await modal.sandboxes.create(app, image, {
 *   experimentalOutboundPolicy: outboundPolicy,
 * });
 * ```
 */
export class ExperimentalOutboundPolicy {
  /** @internal */
  readonly _replacements: readonly HeaderReplacement[];

  /** @internal */
  constructor(_replacements: readonly HeaderReplacement[] = []) {
    this._replacements = _replacements;
  }

  /**
   * Return a new {@link ExperimentalOutboundPolicy} with an added header replacement.
   *
   * @param params.domain Domain the replacements are scoped to. Supports `*.` wildcard
   *   prefixes (matching the apex domain and subdomains) and a bare `"*"`.
   * @param params.headers Header name -> header value. Values support `$KEY` templates
   *   referencing keys in the replacement's `secret`.
   * @param params.secret Named {@link Secret} (e.g. from `client.secrets.fromName`) whose
   *   keys may be referenced in the header value templates. Static replacements pass no secret.
   */
  withHeaderReplacement(params: {
    domain: string;
    headers: Record<string, string>;
    secret?: Secret;
  }): ExperimentalOutboundPolicy {
    const { domain, headers, secret } = params;
    return new ExperimentalOutboundPolicy([
      ...this._replacements,
      { domain, headers: { ...headers }, secret },
    ]);
  }

  /** Check all replacements, throwing `InvalidError` on the first violation found.
   * @internal
   */
  _validate(): void {
    let total = 0;
    for (const replacement of this._replacements) {
      const { domain, headers, secret } = replacement;
      if (domain !== "*" && !DOMAIN_RE.test(domain)) {
        throw new InvalidError(`Invalid domain: ${JSON.stringify(domain)}`);
      }
      const entries = Object.entries(headers);
      if (entries.length === 0) {
        throw new InvalidError("`headers` must contain at least one header");
      }
      for (const [name, value] of entries) {
        if (!HEADER_NAME_RE.test(name)) {
          throw new InvalidError(
            `Invalid header name: ${JSON.stringify(name)}`,
          );
        }
        if (hasInvalidHeaderValueChar(value)) {
          throw new InvalidError(
            `Header value for ${JSON.stringify(name)} must not contain control characters (except tab)`,
          );
        }
        if (secret === undefined && templateReferencesKey(value)) {
          throw new InvalidError(
            `Header value for ${JSON.stringify(name)} references a secret key, but no \`secret\` was passed`,
          );
        }
      }
      if (secret !== undefined && secretEnvDictHydrator(secret) !== undefined) {
        throw new InvalidError(
          "Outbound policies only support named secrets (e.g. `client.secrets.fromName`); " +
            "ephemeral secrets (e.g. `client.secrets.fromObject`) are not yet supported",
        );
      }
      total += entries.length;
    }
    if (total > MAX_HEADER_REPLACEMENTS) {
      throw new InvalidError(
        `Outbound policy cannot have more than ${MAX_HEADER_REPLACEMENTS} header replacements`,
      );
    }
  }

  /** Deduplicated list of secrets referenced by the policy's replacements.
   * @internal
   */
  _secrets(): Secret[] {
    return [
      ...new Set(
        this._replacements.map((s) => s.secret).filter((s) => s !== undefined),
      ),
    ];
  }

  /** Convert to the wire format. Referenced secrets must be hydrated first.
   * @internal
   */
  _toProto(): OutboundPolicyProto {
    return OutboundPolicyProto.create({
      headerReplacements: this._replacements.map((replacement) =>
        OutboundPolicy_HeaderReplacement.create({
          domain: replacement.domain,
          secretId: replacement.secret?.secretId ?? "",
          headers: replacement.headers,
        }),
      ),
    });
  }
}

File: go/outbound_policy.go
package modal

import (
	"fmt"
	"maps"
	"regexp"
	"slices"
	"strings"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
)

var (
	outboundPolicyDomainRe = regexp.MustCompile(
		`^(\*\.)?([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$`,
	)
	outboundPolicyHeaderNameRe = regexp.MustCompile(`^[A-Za-z0-9!#$%&'*+.^_` + "`" + `|~-]+$`)
)

const maxOutboundPolicyHeaderReplacements = 25

func hasInvalidHeaderValueChar(value string) bool {
	return strings.ContainsFunc(value, func(r rune) bool {
		return (r < 0x20 && r != '\t') || r == 0x7F
	})
}

// templateReferencesKey reports whether a header value template references a `$KEY`.
func templateReferencesKey(template string) bool {
	isKeyStart := func(c byte) bool {
		return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
	}
	for i := 0; i+1 < len(template); i++ {
		if template[i] != '$' {
			continue
		}
		c := template[i+1]
		if c == '$' {
			i++ // skip the escaped pair
			continue
		}
		if isKeyStart(c) {
			return true
		}
	}
	return false
}

// ExperimentalOutboundPolicyHeaderReplacement is a single domain-scoped group of header
// replacements.
type ExperimentalOutboundPolicyHeaderReplacement struct {
	// Domain the replacements are scoped to. Supports "*." wildcard prefixes
	// (matching the apex domain and subdomains) and a bare "*".
	Domain string
	// Headers maps header names to values. Values support `$KEY` templates
	// referencing keys in the replacement's Secret. Literal `$` characters are
	// written `$$`.
	Headers map[string]string
	// Secret is a named Secret (e.g. from Client.Secrets.FromName) whose keys
	// may be referenced in the header value templates. Static replacements pass nil.
	Secret *Secret
}

// ExperimentalOutboundPolicy is an immutable configuration for replacing headers in
// outbound HTTPS requests from a Sandbox.
//
// EXPERIMENTAL: the API is subject to change.
//
// Secret values never enter the Sandbox: they are resolved and injected into
// matching requests outside the container.
//
//	secret, _ := client.Secrets.FromName(ctx, "api-token", nil)
//	outboundPolicy := modal.NewExperimentalOutboundPolicy(
//		modal.ExperimentalOutboundPolicyHeaderReplacement{
//			Domain:  "example.com",
//			Secret:  secret,
//			Headers: map[string]string{"Authorization": "Bearer $API_TOKEN"},
//		},
//		modal.ExperimentalOutboundPolicyHeaderReplacement{
//			Domain:  "modal.com",
//			Headers: map[string]string{"X-Trace-Token": "trace_abcd"},
//		},
//	)
type ExperimentalOutboundPolicy struct {
	replacements []ExperimentalOutboundPolicyHeaderReplacement
}

// NewExperimentalOutboundPolicy returns an ExperimentalOutboundPolicy with the given header replacements.
func NewExperimentalOutboundPolicy(replacements ...ExperimentalOutboundPolicyHeaderReplacement) ExperimentalOutboundPolicy {
	p := ExperimentalOutboundPolicy{}
	for _, r := range replacements {
		p = p.WithHeaderReplacement(r)
	}
	return p
}

// WithHeaderReplacement returns a new ExperimentalOutboundPolicy with an added header replacement.
func (p ExperimentalOutboundPolicy) WithHeaderReplacement(replacement ExperimentalOutboundPolicyHeaderReplacement) ExperimentalOutboundPolicy {
	headers := make(map[string]string, len(replacement.Headers))
	maps.Copy(headers, replacement.Headers)
	replacement.Headers = headers
	return ExperimentalOutboundPolicy{replacements: append(slices.Clone(p.replacements), replacement)}
}

// validate checks all replacements, returning the first violation found.
func (p *ExperimentalOutboundPolicy) validate() error {
	if p == nil {
		return nil
	}
	total := 0
	for _, r := range p.replacements {
		if r.Domain != "*" && !outboundPolicyDomainRe.MatchString(r.Domain) {
			return InvalidError{Exception: fmt.Sprintf("Invalid domain: %q", r.Domain)}
		}
		if len(r.Headers) == 0 {
			return InvalidError{Exception: "Headers must contain at least one header"}
		}
		// Sort the names for deterministic error messages and proto conversion.
		for _, name := range slices.Sorted(maps.Keys(r.Headers)) {
			value := r.Headers[name]
			if !outboundPolicyHeaderNameRe.MatchString(name) {
				return InvalidError{Exception: fmt.Sprintf("Invalid header name: %q", name)}
			}
			if hasInvalidHeaderValueChar(value) {
				return InvalidError{Exception: fmt.Sprintf("Header value for %q must not contain control characters (except tab)", name)}
			}
			if r.Secret == nil && templateReferencesKey(value) {
				return InvalidError{Exception: fmt.Sprintf(
					"Header value for %q references a secret key, but no Secret was passed", name)}
			}
		}
		if r.Secret != nil {
			if _, isEnvDict := secretEnvDictHydrator(r.Secret); isEnvDict {
				return InvalidError{Exception: "Outbound policies only support named secrets" +
					" (e.g. Client.Secrets.FromName); ephemeral secrets (e.g. Client.Secrets.FromMap)" +
					" are not yet supported"}
			}
		}
		total += len(r.Headers)
	}
	if total > maxOutboundPolicyHeaderReplacements {
		return InvalidError{Exception: fmt.Sprintf(
			"Outbound policy cannot have more than %d header replacements", maxOutboundPolicyHeaderReplacements)}
	}
	return nil
}

// secrets returns the deduplicated Secrets referenced by the policy's replacements.
func (p *ExperimentalOutboundPolicy) secrets() []*Secret {
	if p == nil {
		return nil
	}
	var out []*Secret
	for _, replacement := range p.replacements {
		if replacement.Secret != nil && !slices.Contains(out, replacement.Secret) {
			out = append(out, replacement.Secret)
		}
	}
	return out
}

// toProto converts to the wire format. Referenced secrets must be hydrated first.
func (p *ExperimentalOutboundPolicy) toProto() *pb.OutboundPolicy {
	if p == nil {
		return nil
	}
	replacements := make([]*pb.OutboundPolicy_HeaderReplacement, 0, len(p.replacements))
	for _, replacement := range p.replacements {
		secretID := ""
		if replacement.Secret != nil {
			secretID = replacement.Secret.SecretID
		}
		replacements = append(replacements, pb.OutboundPolicy_HeaderReplacement_builder{
			Domain:   replacement.Domain,
			SecretId: secretID,
			Headers:  replacement.Headers,
		}.Build())
	}
	return pb.OutboundPolicy_builder{HeaderReplacements: replacements}.Build()
}
